### PR TITLE
[BEAM-409] Ensure precision in division of Longs in ApproximateQuantilesCombineFn

### PR DESCRIPTION
**Pull request by JasonMWhite**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

This PR resolves [BEAM-409](https://issues.apache.org/jira/browse/BEAM-409) by casting a long to a float before division.

@swegner Would you mind reviewing for me?

### CODE REVIEW

**Comment by dhalperi** [comment]
R: @swegner

**Comment by swegner** [comment]
Thanks for this contribution. I'd like to have some unit tests that show the impact of this bug and to verify your fix. I've opened PR #856 with unit tests. Once that PR looks good, would you mind merging into yours to submit?

**Comment by JasonMWhite** [comment]
Thanks @swegner, your changes look good to me and I've cherry-picked them into #853. I learned quite a bit while reading it, thanks :)

**Comment by JasonMWhite** [comment]
`mvn clean verify` runs fine for me, and the failures on Jenkins and Travis look unrelated to this PR, or to each other for that matter. I've pulled the commits from #856 into this PR.

**Comment by swegner** [comment]
LGTM

R: @dhalperi for committer approval and merge.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ApproximateQuantilesTest.java [hunk 1]
```diff
@@ -35,283 +37,385 @@
...
-  private PCollection<Integer> intRangeCollection(Pipeline p, int size) {
-    return p.apply("CreateIntsUpTo(" + size + ")", Create.of(intRange(size)));
-  }
+      this.combineFn = ApproximateQuantilesCombineFn.create(
+          10, new Top.Largest<Long>(), maxInputSize, epsilon);
+    }
+
+    @Before
+    public void setUp() {
+      LOG.info("Epsilon: {}, Max input size: {}", epsilon, maxInputSize);
```
Comment (dhalperi) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/ApproximateQuantilesTest.java]: please drop. Instead, you can do this directly in the `@Parameters` annotation, then you also see which parameters failed on the command-line (or in your IDE) when it does.

Example: https://github.com/apache/incubator-beam/blob/064796dcc6bffe5028411006c196129ee8766c44/sdks/java/core/src/test/java/org/apache/beam/sdk/io/range/ByteKeyRangeEstimateFractionTest.java#L38)

**Comment by JasonMWhite** [comment]
On vacation ATM, will get back to it when I return in a week.

**Comment by dhalperi** [comment]
sgtm.

**Comment by dhalperi** [comment]
Hi Jason, any updates here?

**Comment by JasonMWhite** [comment]
I'll close for now and reopen later. Realized I needed to up my Java skills a bit first. :)

**Pull request closed without merging**